src/libcore/pin.rs

```diff
@@ -94,6 +95,106 @@
+//! the question arises whether `fn(Pin<&[mut] Struct>) -> Pin<&[mut] Field>`
+//! is an operation that can be soundly added to the API.
+//!
+//! This question is closely related to the question of whether pinning is "structural":
+//! when you have pinned a container, have you pinned its contents? Adding a
+//! projection to the API answers that question with a "yes" by offering pinned access
+//! to the contents.
+//!
+//! In general, as the author of a type you get to decide whether pinning is structural, and
+//! whether projections are provided. However, there are a couple requirements to be
+//! upheld when adding projection operations:
+//!
+//! 1. The container must only be [`Unpin`] if all its fields are `Unpin`. This is the default,
+//!    but `Unpin` is a safe trait, so as the author of the container it is your responsibility
+//!    *not* to add something like `impl<T> Unpin for Container<T>`. (Notice that adding a
+//!    projection operation requires unsafe code, so the fact that `Unpin` is a safe trait
+//!    does not break the principle that you only have to worry about any of this if
+//!    you use `unsafe`.)
+//! 2. The destructor of the container must not move out of its argument. This is the exact
+//!    point that was raised in the [previous section][drop-impl]: `drop` takes `&mut self`,
+//!    but the container (and hence its fields) might have been pinned before.
+//!    You have to guarantee that you do not move a field inside your `Drop` implementation.
+//! 3. Your container type must *not* be `#[repr(packed)]`. Packed structs have their fields
+//!    moved around when they are dropped to properly align them, which is in conflict with
+//!    claiming that the fields are pinned when your struct is.
+//! 4. You must make sure that you uphold the [`Drop` guarantee][drop-guarantee]:
+//!    you must make sure that, once your container is pinned, the memory containing the
+//!    content is not overwritten or deallocated without calling the content's destructors.
+//!    This can be tricky, as witnessed by `VecDeque`: the destructor of `VecDeque` can fail
+//!    to call `drop` on all elements if one of the destructors panics. This violates the
+//!    `Drop` guarantee, because it can lead to elements being deallocated without
+//!    their destructor being called.
+//! 5. You must not offer any other operations that could lead to data being moved out of
+//!    the fields when your type is pinned. This is usually not a concern, but can become
+//!    tricky when interior mutability is involved. For example, imagine `RefCell`
+//!    would have a method `fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut T>`.
+//!    This would be catastrophic, because it is possible to move out of a pinned
+//!    `RefCell`: from `x: Pin<&mut RefCell<T>>`, use `let y = x.into_ref().get_ref()` to obtain
+//!    `y: &RefCell<T>`, and from there use `y.borrow_mut().deref_mut()` to obtain `&mut T`
```

Comment:
How does this tie in to the `get_pin_mut` mentioned above? I don't see it being used here?

Reply:
I expanded the explanation.